src/orange/universes/SimpleUnitTracker.hh

```diff
@@ -113,5 +125,182 @@ CELER_FUNCTION auto SimpleUnitTracker::initialize(LocalState state) const
+    }
+
+    if (is_simple)
+    {
+        // No interior surfaces: closest distance is next boundary
+        return this->simple_intersect(state, vol);
+    }
+    else
+    {
+        return this->complex_intersect(state, vol);
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate distance to the next boundary for nonreentrant cells.
+ */
+CELER_FUNCTION auto
+SimpleUnitTracker::simple_intersect(LocalState state, VolumeView vol) const
+    -> Intersection
+{
+    CELER_EXPECT(state.temp_next && vol.num_intersections() > 0);
+
+    // Crossing any surface will leave the cell; perform a linear search for
+    // the smallest (but nonzero) distance
+    size_type distance_idx;
+    {
+        const real_type* distance_ptr = celeritas::min_element(
+            state.temp_next.distance,
+            state.temp_next.distance + state.temp_next.size,
+            detail::CloserNonzeroDistance{});
+        CELER_ASSERT(*distance_ptr > 0);
+        distance_idx = distance_ptr - state.temp_next.distance;
+    }
+
+    if (state.temp_next.distance[distance_idx] == no_intersection())
+        return {};
+
+    // Determine the crossing surface
+    SurfaceId surface;
+    {
+        FaceId face = state.temp_next.face[distance_idx];
+        CELER_ASSERT(face);
+        surface = vol.get_surface(face);
+        CELER_ASSERT(surface);
+    }
+
+    Sense cur_sense;
+    if (surface == state.surface.id())
+    {
+        // Crossing the same surface that we're currently on and inside (e.g.
+        // on the inside surface of a sphere, and the next intersection is the
+        // other side)
+        cur_sense = state.surface.sense();
+    }
+    else
+    {
+        auto calc_sense = make_surface_action(Surfaces{params_.surfaces},
+                                              detail::CalcSense{state.pos});
+
+        SignedSense ss = calc_sense(surface);
+        CELER_ASSERT(ss != SignedSense::on);
+        cur_sense = to_sense(ss);
+    }
+
+    // Post-surface sense will be on the other side of the surface
+    Intersection result;
+    result.surface  = {surface, flip_sense(cur_sense)};
+    result.distance = state.temp_next.distance[distance_idx];
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate boundary distance if internal surfaces are present.
+ *
+ * In "complex" cells, crossing a surface can still leave the particle in an
+ * "inside" state.
+ *
+ * We have to iteratively track through all surfaces, in order of minimum
+ * distance, to determine whether crossing them in sequence will cause us to
+ * exit the cell.
+ */
+CELER_FUNCTION auto
+SimpleUnitTracker::complex_intersect(LocalState state, VolumeView vol) const
+    -> Intersection
+{
+    // Partition intersections (enumerated from 0 as the `idx` array) into
+    // valid (finite nonzero) and invalid (infinite-or-zero) groups.
```

Comment:
Yep indeed. I think I wrote the comment that way because all of the surface "distance to boundary" calculations return strictly nonnegative distances and I just needed to exclude the zeros. I'll update the comment.

Now that you mention it though, I could probably just update the quadratic solver to not return zero-distance results. I didn't do this earlier because earlier in the package's evolution I was still considering allowing zero-distance movements as long as it changed the logical state.